### PR TITLE
Brittany and Maddie

### PR DESCRIPTION
**Pull request by brittanyrjones**
# Scrabble
Congratulations! You're submitting your assignment.

## Comprehension Questions

Feature | Feedback
--- | ---
What are the advantages of using git when collaboratively working on one code base? |  The advantages of using git to collaborate: Two people can work on the code at the same time, and compare their changes. Another advantage, is that one person can make a change in the code, and the other person can pull those changes, and continue to work on the code. This is especially helpful when there is a driver, and the driver role switches. 
Why do you think we chose to make the `score` method in the `Scoring` class a **class method** instead of an **instance method**? | The score method in the scoring class is a class method instead of an instance method because it will need to be accessed by each player. each time there is a new instance of player, we will not have to redefine a scoring method.
Describe an example of a test you wrote for a _nominal case_. | One example of a nominal test, was if there were a total of 98 tiles, then you could not continue to draw tiles if there were already 98 drawn. 
Describe an example of a test you wrote for an _edge case_. | An edge case that we tested was if there were no tiles drawn at all, and if there were more than 98 tiles drawn (5000 tiles). 
Does your code use any methods from the `Enumerable` mixin? If so, where and why was it helpful? | Yes, one place we used it was "if max_score_words.any?  |word| word.length == 7"Where we wanted to iterate through each word to see if any of the words were 7 letters long. This made the code easier to follow, and dried up the method which would have been much longer. 
What was one method you and your pair used to debug code? | One method we used was writing a test to fail, (drawing tiles) and then looking at the stack trace to see if the code was breaking exactly where we were expecting it to break. Another method we tried to use (not as successful as hoping) was using the pry method to stop our code execution where we wanted to input our own numbers to see how the code was running. 
Is there a specific piece of code you'd like feedback on? | We had a lot of issues with the player class. We wanted to make sure we worked on this together, because when we came together to talk about our solutions, it seemed that we came up with better solutions when we worked together, and those solutions made sense both of us. When we worked on the code alone, and then came back together to show each other our ideas, it just got a little confusing for us. 
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? | Feedback: We felt that we worked really well as a team. One barrier for us was not having more time to meld and work out our ideas face to face. It was kind of hard for us to communicate our ideas just through slack and git, as we liked having the whiteboards, and real time demonstration from our partner on their idea. We are still working to complete the player class together.

### CODE REVIEW

File: lib/scoring.rb [hunk 1]
```diff
@@ -1,9 +1,81 @@
...
+        # First part is splitting the hash into an array of the
+        # key values 'A-Z' and then we are checking if any characters
+        # are outside that range (This includes blank or nothing).
+        # If so, automatically return nil
+        #
+        # Else, if all the characters are valid letters, take the value
+        # associated with the key and put that in an array of integers
+
+
+        if !character_check.include?(character) || letters.length == 0 || letters.length > 7
```
Comment (CheezItMan) [inline lib/scoring.rb:34]: This `if` doesn't need to be in the loop, at least the checks for the length of the array.

File: lib/scoring.rb [hunk 2]
```diff
@@ -1,9 +1,81 @@
...
+          return nil
+
+        else
+          values_of_letters << LETTER_VALUES.fetch(character)
```
Comment (CheezItMan) [inline lib/scoring.rb:38]: Instead of using `fetch` you could just

```ruby
values_of_letters << LETTER_VALUES[character]
```

File: specs/scoring_spec.rb [hunk 3]
```diff
@@ -9,54 +9,66 @@
...
     end
 
     it 'returns the highest word if there are two words' do
+      array_of_words = ["zebra", "zebras"]
+      Scrabble::Scoring.highest_score_from(array_of_words).must_equal("zebras")
     end
 
     it 'if tied, prefer a word with 7 letters' do
+      array_of_words = ["ww", "bananas"]
+      Scrabble::Scoring.highest_score_from(array_of_words).must_equal("bananas")
```
Comment (CheezItMan) [inline specs/scoring_spec.rb:61]: It's also important to vary the order here, just in case order played a role.

File: specs/scoring_spec.rb [hunk 4]
```diff
@@ -9,54 +9,66 @@
...
     end
 
     it 'if tied and no word has 7 letters, prefers the word with fewer letters' do
+      array_of_words = ["x", "brasas"]
+      Scrabble::Scoring.highest_score_from(array_of_words).must_equal("x")
```
Comment (CheezItMan) [inline specs/scoring_spec.rb:66]: Same as above

File: lib/player.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      @plays = []
+      # #plays: returns an Array of the words played by the player
+      @total_score = 0
+      @tiles=[]
+
+    end
+
+    def winner
+      return @total_score > 100
+      end
```
Comment (CheezItMan) [inline lib/player.rb:34]: **Fix** your indentation!

Also this method should be named: `won?`

File: lib/player.rb [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    def play(word)
+      if winner == false
+        @plays << word
+        @total_score += Scoring.score(word)
+      else
+        return false
+      end
+    end
+
+
```
Comment (CheezItMan) [inline lib/player.rb:45]: `highest_word_score`???

`highest_scoring_word`????

File: lib/player.rb [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+      @tiles=[]
+
+    end
+
+    def winner
+      return @total_score > 100
+      end
+
+    def play(word)
+      if winner == false
```
Comment (CheezItMan) [inline lib/player.rb:37]: This method is supposed to return the score of the word played, or false if the player has already won.

File: lib/tilebag.rb [hunk 8]
```diff
@@ -0,0 +1,52 @@
+module Scrabble
+  class TileBag
+    #player can see letters
+    attr_reader :letter_quantity
+
+
+    # Creating an instance of letter quantity to be used though class tilebag.
+    def initialize
+      @default_set = {A:9, B:2, C:2, D:4, E:12, F:2, G:3, H:2, I:9, J:1, K:1, L:4, M:2, N:6, O:8, P:2, Q:1, R:6, S:4, T:6, U:4, V:2, W:2, X:1, Y:2, Z:1}
```
Comment (CheezItMan) [inline lib/tilebag.rb:9]: There's a problem with this structure.  It gives each letter an **equal** chance of being drawn despite the fact that there are 2 "M" tiles and 9 "A" tiles.  Uh oh!

**Comment by CheezItMan** [comment]
# Scrabble
## What We're Looking For

Feature | Feedback
--- | ---
**General** |
Answered comprehension questions | Check, we made `score` and `highest_score_from` class methods because those actions don't make a lot of sense as a `Player` action.  
Both Teammates contributed to the codebase | Both teammate put in commits!
Regular Commits with meaningful commit messages |  Good commit messages and a good number of commits
**Scoring class** |
`score` method |  Check
Uses appropriate data structure to store the letter score | Check
Appropriately handles edge cases |  Check
`highest_score_from` method |  Check
Appropriately handles edge cases |  Check 
Test implementations for `highest_score_from`| Check, but also test for varied order in the words.
**Player class** |
Uses an array to keep track of words played | Check
Uses existing code to create `highest_scoring_word` and `highest_word_score` | **MISSING**
Returns `true` or `false` from `won?` method |  the method is named wrong.  
Tests edge cases on the `play(word)` |  Missed the return value.  **MISSING TESTS**
**TileBag class** |
Uses a data structure to keep the set of default tiles in the bag |  See my notes on this.  The hash is problematic.  
`draw_tiles` method uses the parameter to draw a set of random tiles | It works, but players are more likely to get a "Z" than they should.  
`tiles_remaining` method returns a count |  Check
Tests `draw_tiles` edge cases |  Check
Summary | Just a note, please make sure that any code you submit has proper indentation.  It looks like you missed big sections of Wave 2.  Maybe you ran out of time a bit?  You need to work on testing of course and creating and using instance methods.
